crates/qasm2/src/lex.rs

```diff
@@ -0,0 +1,772 @@
+            return match self.next_byte()? {
+                // In the case of a float that begins with '.', we require at least one digit, so
+                // just force consume it and then loop over the rest.
+                Some(b'0'..=b'9') => {
+                    while let Some(b'0'..=b'9') = self.peek_byte()? {
+                        self.next_byte()?;
+                    }
+                    self.lex_float_exponent(start_col)
+                }
+                _ => Err(QASM2ParseError::new_err(message_generic(
+                    Some(&Position::new(&self.filename, self.line, start_col)),
+                    "expected a numeric fractional part after the bare decimal point",
+                ))),
+            };
+        }
+        while let Some(b'0'..=b'9') = self.peek_byte()? {
+            self.next_byte()?;
+        }
+        match self.peek_byte()? {
+            Some(b'.') => {
+                self.next_byte()?;
+                while let Some(b'0'..=b'9') = self.peek_byte()? {
+                    self.next_byte()?;
+                }
+                return self.lex_float_exponent(start_col);
+            }
+            // In this situation, what we've lexed so far is an integer (maybe with leading
+            // zeroes), but it can still be a float if it's followed by an exponent.  This
+            // particular path is not technically within the spec (so should be subject to `strict`
+            // mode), but pragmatically that's more just a nuisance for OQ2 generators, since many
+            // languages will happily spit out something like `5e-5` when formatting floats.
+            Some(b'e' | b'E') => {
+                return if self.strict {
+                    Err(QASM2ParseError::new_err(message_generic(
+                        Some(&Position::new(&self.filename, self.line, start_col)),
+                        "[strict] all floats must include a decimal point",
+                    )))
+                } else {
+                    self.lex_float_exponent(start_col)
+                }
+            }
+            _ => (),
+        }
+        if first == b'0' && self.col - start_col > 1 {
+            // Integers can't start with a leading zero unless they are only the single '0', but we
+            // didn't see a decimal point.
+            Err(QASM2ParseError::new_err(message_generic(
+                Some(&Position::new(&self.filename, self.line, start_col)),
+                "integers cannot have leading zeroes",
+            )))
+        } else {
+            self.expect_word_boundary("an integer", start_col)?;
+            Ok(TokenType::Integer)
+        }
+    }
+
+    /// Lex a text-like token into a complete token.  This can return any of the keyword-like
+    /// tokens (e.g. [TokenType::Pi]), or a [TokenType::Id] if the token is not a built-in keyword.
+    fn lex_textlike(&mut self, start_col: usize) -> PyResult<TokenType> {
+        let first = self.line_buffer[start_col];
```

Comment:
Hmm yeah, it seems to me like `start_col` is just `self.col` at the start of the function. Do we need a parameter for that, or can we just get it from `self`?

Reply:
The `lex_*` functions are all continuation lexers, so in `textlike` and `numeric` I think it's `self.col - 1`, and in `float_exponent` it's completely unknown, so definitely needs passing.  I didn't want to have the magic `self.col - 1` floating around, and I needed the passing for `float_exponent`, so I think I used `start_col` in all of them for consistency.

I can change the fixed ones to `self.col - 1` if you much prefer.